### PR TITLE
use common sql interfaces

### PR DESCRIPTION
**Pull request by prolic**
See also here: https://github.com/prolic/sql
(We would need to move the repo to amphp organization)

If I missed something, let me know please.

### CODE REVIEW

**Comment by prolic** [comment]
see also https://github.com/amphp/mysql/pull/74

**Comment by prolic** [comment]
@trowski the pool interface is here: https://github.com/amphp/sql/blob/master/src/Pool.php

this conflicts with your latest commit, can you revert it?

**Comment by prolic** [comment]
@trowski updated as well

File: src/Internal/PooledStatement.php [hunk 1]
```diff
@@ -3,14 +3,15 @@
...
 use Amp\Postgres\Pool;
-use Amp\Postgres\Statement;
 use Amp\Promise;
+use Amp\Sql\Operation;
+use Amp\Sql\Statement;
 use function Amp\call;
 
-final class PooledStatement implements Statement {
-    /** @var \Amp\Postgres\Pool */
+final class PooledStatement implements Statement
```
Comment (trowski) [inline src/Internal/PooledStatement.php]: This class is no longer needed.

File: src/Executor.php [hunk 2]
```diff
@@ -4,51 +4,52 @@
 
 use Amp\Promise;
 
-interface Executor {
```
Comment (trowski) [inline src/Executor.php]: Should extend `Amp\Sql\Executor` and not repeat the methods in that interface.

File: src/Link.php [hunk 3]
```diff
@@ -3,27 +3,18 @@
 namespace Amp\Postgres;
 
 use Amp\Promise;
+use Amp\Sql\Link as SqlLink;
 
-interface Link extends Executor {
```
Comment (trowski) [inline src/Link.php]: This interface still needs to extend Executor as well.

File: src/Pool.php [hunk 4]
```diff
@@ -3,43 +3,9 @@
...
-     */
-    public function getIdleTimeout(): int;
-
-    /**
-     * @param int $timeout Number of seconds a connection may remain idle before it is automatically closed.
-     */
-    public function setIdleTimeout(int $timeout);
+interface Pool extends Link, SqlPool
+{
+    public function notify(string $channel, string $payload = ""): Promise;
```
Comment (trowski) [inline src/Pool.php]: This method is already defined in `Executor`.

**Comment by prolic** [comment]
@trowski all done

**Comment by prolic** [comment]
@trowski there are still some tests failing, wanna look into it? you can submit a PR to my branch.

**Comment by DaveRandom** [comment]
@prolic I have fixed the failing tests in the `fix/sql-interfaces-pr` branch.

Rather than PR'ing this against your fork and creating a superfluous merge commit, you should be able to simply pull the changes from that branch into yours by doing something like:

```
# checkout the sql branch locally
git checkout sql

# pull the changes from the branch with the fixes
git pull git@github.com:amphp/postgres.git fix/sql-interfaces-pr

# push the changes up to your fork
git push origin sql
```

**Comment by prolic** [comment]
@DaveRandom thanks, it's done.

**Comment by DaveRandom** [comment]
As it currently stands, I don't think the interface layout is quite right.

The thing that sets off alarm bells is that neither of the concrete `Handle` classes implement `Link`, and thus their `listen()` methods are not covered by an interface. It's not as simple as changing the `implements` list, as they currently do not implement `transaction()`.

I also have some issues with the interfaces defined upstream by the `Sql` package, I will open an issue there and link it here.

**Comment by trowski** [comment]
@DaveRandom Yep, looks like `listen()` should have been defined in `handle()` instead of `Link`. This means `Transaction` would need to add `listen()`. Should it? If so, should listening end when the transaction is committed/rolled-back? If not, we may need yet another interface.

`Handle` also does not need to define `lastUsedAt()`, as it is defined in `Executor` in amphp/sql.

**Comment by trowski** [comment]
@DaveRandom I made some changes based on your feedback in the [squash branch](https://github.com/amphp/postgres/compare/squash). Have a look at the commits I made today and let me know what you think. /cc @prolic

**Comment by DaveRandom** [comment]
@trowski that seems like a reasonable change, yes. I'm a little surprised tbh, I was expecting `LISTEN` to be a link-level state, but [docs](https://www.postgresql.org/docs/current/static/sql-listen.html) indicate that doing it within a transaction will alter the semantics of the command to reflect the transaction, so it seems that both `listen()` and `notify()` should always be available together.

**Comment by trowski** [comment]
@DaveRandom It sounds like the listen command doesn't take affect until after the transaction is committed? I can do that too I guess… but I will have to keep the connection busy while the listener is active (i.e., the `Transaction` object will stick around in memory, not releasing the associated connection back to pool). That would probably be expected behavior based on the docs.

**Comment by DaveRandom** [comment]
@trowski pragmatically, it might be better to just not support listen/notify in transactions, at least for now, Using them in a transaction is functionally identical to changing the logic such that they are called on successful commit, which seems easy to accomplish in app code and avoids what seems like a lot of complexity in the library code to make this interact sanely with a pool.

Doing so initially would not preclude adding the functionality later in a backwards-compatible, if the interface structure is done right,

Just a thought, possibly not a great one :-P

**Comment by prolic** [comment]
I think this can be closed now? Will you update mysql as well, soon?

**Comment by kelunik** [comment]
@prolic I can close this now without merging, right?

**Comment by prolic** [comment]
Yes, it's already merged with rebased branch.

**Pull request closed without merging**